js/ui/windowManager.js

```diff
@@ -1832,3 +1887,608 @@ const WindowManager = new Lang.Class({
+        session.firstFrameConnection = firstFrameConnection;
+        this._firstFrameConnections.push(firstFrameConnection);
+    },
+
+    _windowAppPositionChanged: function(window, session) {
+        let rect = session.actorApp.meta_window.get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorBuilder)
+            return;
+        session.actorBuilder.meta_window.move_resize_frame(false,
+                                                           rect.x,
+                                                           rect.y,
+                                                           rect.width,
+                                                           rect.height);
+    },
+
+    _windowAppSizeChanged: function(window, session) {
+        let rect = session.actorApp.meta_window.get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorBuilder)
+            return;
+        session.actorBuilder.meta_window.move_resize_frame(false,
+                                                           rect.x,
+                                                           rect.y,
+                                                           rect.width,
+                                                           rect.height);
+    },
+
+    _windowBuilderPositionChanged: function(window, session) {
+        let rect = session.actorBuilder.meta_window.get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorApp)
+            return;
+        session.actorApp.meta_window.move_resize_frame(false,
+                                                       rect.x,
+                                                       rect.y,
+                                                       rect.width,
+                                                       rect.height);
+    },
+
+    _windowBuilderSizeChanged: function(window, session) {
+        let rect = session.actorBuilder.meta_window.get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorApp)
+            return;
+        session.actorApp.meta_window.move_resize_frame(false,
+                                                       rect.x,
+                                                       rect.y,
+                                                       rect.width,
+                                                       rect.height);
+    },
+
+    _windowMinimized: function(shellwm, actor, session) {
+        // take the actor that we minimized and emit a minimized
+        // signal for it, though setting a flag internally so that
+        // we don't re-enter this signal handler.
```

Comment:
Is `minimize()` not a no-op if the window is already minimized? I am thinking that you may get away without the additional state here if so.